service_discovery/core/src/service_registry_impl.rs

```diff
@@ -0,0 +1,420 @@
+                None => {
+                    lock.insert(service_identifier.clone(), service_to_register.clone());
+                    info!(
+                        "Registered new service in the service registry: {service_to_register:?}"
+                    );
+                    let register_response = RegisterResponse {
+                        registration_status: RegistrationStatus::NewlyRegistered as i32,
+                    };
+                    Ok(Response::new(register_response))
+                }
+            }
+        }
+    }
+
+    /// Unregisters a service by removing it from the registry.
+    ///
+    /// # Arguments
+    ///
+    /// * `request` - Contains the service identification information for the service to unregister
+    async fn unregister(
+        &self,
+        request: Request<UnregisterRequest>,
+    ) -> Result<Response<UnregisterResponse>, Status> {
+        let request_inner = request.into_inner();
+        let service_identifier = ServiceIdentifier {
+            namespace: request_inner.namespace,
+            name: request_inner.name,
+            version: request_inner.version,
+        };
+        info!("Received an unregister request for: {service_identifier:?}");
+
+        let mut lock = self.registry_map.write();
+        match lock.remove(&service_identifier) {
+            Some(removed_service) => {
+                info!(
+                    "Successfully removed service entry from service registry: {removed_service:?}"
+                );
+                Ok(Response::new(UnregisterResponse {}))
+            }
+            None => {
+                let not_found_message =
+                    format!("Unable to remove service from registry: {service_identifier:?}");
+                warn!(not_found_message);
+                Err(Status::not_found(not_found_message))
+            }
+        }
+    }
+
+    /// Discovers a list of services based on the namespace, or logical grouping of services.
+    ///
+    /// # Arguments
+    ///
+    /// * `request` - Contains the namespace of the services to be discovered
+    async fn discover_by_namespace(
+        &self,
+        request: Request<DiscoverByNamespaceRequest>,
+    ) -> Result<Response<DiscoverByNamespaceResponse>, Status> {
+        let namespace = request.into_inner().namespace;
+
+        // This block controls the lifetime of the lock.
+        let service_list: Vec<ServiceMetadata> = {
+            let lock = self.registry_map.read();
+            lock.iter()
+                .filter_map(|(service_identifier, service_metadata)| {
+                    if service_identifier.namespace == namespace {
+                        Some(service_metadata.clone())
+                    } else {
+                        None
+                    }
+                })
+                .collect()
+        };
+        if service_list.is_empty() {
+            Err(Status::not_found(format!("No registrations found for namespace {namespace}")))
```

Comment:
do we consider this an error or should we just return an empty list?

Reply:
This is a good question. For discovering a single service, we fail if it's not found. But since in this case, the caller would need to consume the list to decide what to do anyway, I will change it to just return an empty list